libraries/AP_DDS/AP_DDS_Client.cpp

```diff
@@ -582,22 +597,81 @@ void AP_DDS_Client::on_request(uxrSession* uxr_session, uxrObjectId object_id, u
  */
 void AP_DDS_Client::main_loop(void)
 {
-    if (!init() || !create()) {
-        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "%s Creation Requests failed", msg_prefix);
+    if (!init_transport()) {
         return;
     }
-    GCS_SEND_TEXT(MAV_SEVERITY_INFO, "%s Initialization passed", msg_prefix);
-
-    populate_static_transforms(tx_static_transforms_topic);
-    write_static_transforms();
 
+    //! @todo check for request to stop task
     while (true) {
-        hal.scheduler->delay(1);
-        update();
+        if (comm == nullptr) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "DDS Client: transport invalid, exiting");
+            return;
+        }
+
+        // check ping
+        const uint64_t ping_timeout_ms{1000};
+        const uint8_t ping_max_attempts{10};
+        if (!uxr_ping_agent_attempts(comm, ping_timeout_ms, ping_max_attempts)) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "DDS Client: No ping response, exiting");
+            return;
+        }
+
+        // create session
+        if (!init_session() || !create()) {
+            GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "DDS Client: Creation Requests failed");
+            return;
+        }
+        connected = true;
+        GCS_SEND_TEXT(MAV_SEVERITY_INFO, "DDS Client: Initialization passed");
+
+        populate_static_transforms(tx_static_transforms_topic);
+        write_static_transforms();
+
+        uint64_t last_ping_ms{0};
+        uint8_t num_pings_missed{0};
+        bool had_ping_reply{false};
+        while (connected) {
+            hal.scheduler->delay(1);
+
+            // publish topics
+            update();
+
+            // check ping response
+            if (session.on_pong_flag == 1) {
+                had_ping_reply = true;
+            }
+
+            const auto cur_time_ms = AP_HAL::millis64();
+            if (cur_time_ms - last_ping_ms > DELAY_PING_MS) {
+                last_ping_ms = cur_time_ms;
+
+                if (had_ping_reply) {
+                    num_pings_missed = 0;
+
+                } else {
+                    ++num_pings_missed;
+                }
+
+                uxr_ping_agent_session(&session, 0, 1);
```

Comment:
Do you intend to set a timeout of 0 here? Note - for a non-obvious C-api like this one, I might actually just create some constexpr locals to show what is being passed in. 
```
constexpr int timout_ping_ms {0};
constexpr int ping_attempts {1};
uxr_ping_agent_session(&session, timout_ping_ms, ping_attempts);
```
uxr_ping_agent_session(&session, 0, 1);

Reply:
Added const stack variables in:
https://github.com/ArduPilot/ardupilot/pull/25228/commits/a14e0af23432732532e703e0b75dd02cc4cf618c

Didn't add constexpr as we're adding for readability and are happy for the compiler to optimise these away.

The 0 timeout looks odd doesn't it? This is consistent with the approach used by the PX4 DDS implementation - it must be trying to poll at least once otherwise I'd expect a timeout. Need to trace this more thoroughly to understand why it does not fail every time.